File: minimumheap.py

import sys
class Maximum_Heap:
	def __init__(self, maxsize):
		self.maxsize = maxsize
		self.size = 0
		self.ArrayOfHeap = [0] * (self.maxsize + 1)
		self.ArrayOfHeap[0] = sys.maxsize
		self.FRONT = 1

	def Node_parent(self, position):
		return position-1 // 2

	def Node_leftChild(self, position):
		return 2 * position

	def Node_rightChild(self, position):
		return (2 * position) + 1

	def CheckLeafFunction(self, position):

		if position >= (self.size//2) and position <= self.size:
			return True
		return False

	def SwapFunction(self, first_position, second_position):
		self.ArrayOfHeap[first_position], self.ArrayOfHeap[second_position] = (self.ArrayOfHeap[second_position],self.ArrayOfHeap[first_position])

	def FindMaxHeap(self, position):
		if not self.CheckLeafFunction(position):
			if (self.ArrayOfHeap[position] < self.ArrayOfHeap[self.Node_leftChild(position)] or
				self.ArrayOfHeap[position] < self.ArrayOfHeap[self.Node_rightChild(position)]):
				if (self.ArrayOfHeap[self.Node_leftChild(position)] >
					self.ArrayOfHeap[self.Node_rightChild(position)]):
					self.SwapFunction(position, self.Node_leftChild(position))
					self.FindMaxHeap(self.Node_leftChild(position))

				else:
					self.SwapFunction(position, self.Node_rightChild(position))
					self.FindMaxHeap(self.Node_rightChild(position))

	def ElementInsert(self, value):
		if self.size >= self.maxsize:
			return
		self.size += 1
		self.ArrayOfHeap[self.size] =value

		current = self.size

		while (self.ArrayOfHeap[current] < self.ArrayOfHeap[self.Node_parent(current)]):
			self.SwapFunction(current, self.Node_parent(current))
			current = self.Node_parent(current)

	def DisplayFunction(self):
		for x in range(1, (self.size // 2) + 1):
			print(" Parent Node : " + str(self.ArrayOfHeap[x]))
			print("Left Child : " + str(self.ArrayOfHeap[2 * x]) +
			" Right Child : " + str(self.ArrayOfHeap[2 * x + 1]))

	def extractMaximuHeap(self):
		remove = self.ArrayOfHeap[self.FRONT]
		self.ArrayOfHeap[self.FRONT] = self.ArrayOfHeap[self.size]
		self.size -= 1
		self.FindMaxHeap(self.FRONT)

		return remove
if __name__ == "__main__":

	print('The maxHeap is ')

	m = Maximum_Heap(15)
	m.ElementInsert(1)
	m.ElementInsert(2)
	m.ElementInsert(3)
	m.ElementInsert(4)
	m.ElementInsert(5)
	m.ElementInsert(6)
	m.ElementInsert(7)
	m.ElementInsert(8)
	m.ElementInsert(9)

	m.DisplayFunction()

	print("The Max val is " + str(m.extractMaximuHeap()))

File: minimum heap.cpp
#include<iostream>
using namespace std;
int ArrayOfHeap[100];
int size;
int Size_Of_Maximum;
void MaximumHeapFunction(int n)
{

		Size_Of_Maximum = n;
		size = 0;
}
int Node_Parent(int position) { return (position - 1) / 2; }
int Node_leftChild(int position) { return (2 * position) + 1; }
int Node_rightChild(int position){ return (2 * position) + 2; }

bool CheckLeafFunction(int position)
{
    if (position > (size / 2) && position <= size)
        {
			return true;
		}
		return false;
}

void SwapFunction(int FirstPosition, int SecondPosition)
{
		int a;
		a = ArrayOfHeap[FirstPosition];
		ArrayOfHeap[FirstPosition] = ArrayOfHeap[SecondPosition];
		ArrayOfHeap[SecondPosition] = a;
}

void FindMaxHeap(int position)
{
		if (CheckLeafFunction(position))
			return;
		if (ArrayOfHeap[position] < ArrayOfHeap[Node_leftChild(position)] || ArrayOfHeap[position] < ArrayOfHeap[Node_rightChild(position)])
        {

			if (ArrayOfHeap[Node_leftChild(position)]> ArrayOfHeap[Node_rightChild(position)])
            {
				SwapFunction(position, Node_leftChild(position));
				FindMaxHeap(Node_leftChild(position));
			}
			else
			{
				SwapFunction(position, Node_rightChild(position));
				FindMaxHeap(Node_rightChild(position));
			}
		}
}
void ElementInsert(int value)
{
		ArrayOfHeap[size] = value;

		int current = size;
		while (ArrayOfHeap[current] < ArrayOfHeap[Node_Parent(current)])
        {
			SwapFunction(current, Node_Parent(current));
			current = Node_Parent(current);
		}
		size++;
}

void DisplayFunction()
{
	for(int i=0;i<size/2;i++)
    {

			cout<<"Node_Parent Node : "<<ArrayOfHeap[i]<<"\n";

			if(Node_leftChild(i)<size)
			cout<<" Left Child Node: " <<ArrayOfHeap[Node_leftChild(i)]<<"\n";

			if(Node_rightChild(i)<size)
				cout<<" Right Child Node: "<<ArrayOfHeap[Node_rightChild(i)]<<"\n";

				cout<<"\n";

    }

}
int extractMaximuHeap()
{
		int remove = ArrayOfHeap[0];
		ArrayOfHeap[0] = ArrayOfHeap[--size];
		FindMaxHeap(0);
		return remove;
}

main()
{

        MaximumHeapFunction(15);
        ElementInsert(1);
		ElementInsert(2);
		ElementInsert(3);
		ElementInsert(4);
		ElementInsert(5);
		ElementInsert(6);
		ElementInsert(7);
		ElementInsert(8);
		ElementInsert(9);
		DisplayFunction();
		cout<<"The maximum  value in heap "<<extractMaximuHeap()<<"\n";
}
